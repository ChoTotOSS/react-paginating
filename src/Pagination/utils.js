export function getRange(start, end) {
  return [...Array(end - start + 1)].map((_, i) => start + i);
}

export function getPageInfo({ limit, pageCount, total, page }) {
  const totalPages = Math.ceil(total / limit);

  const totalResults = parseInt(total, 10);
  let currentPage = parseInt(page, 10);

  if (currentPage < 1) {
    currentPage = 1;
  }

  if (currentPage > totalPages) {
    currentPage = totalPages;
  }

  let firstPage = Math.max(1, currentPage - Math.floor(pageCount / 2));

  let lastPage = Math.min(totalPages, currentPage + Math.floor(pageCount / 2));

  if (lastPage - firstPage + 1 < pageCount) {
    if (currentPage < totalPages / 2) {
      lastPage = Math.min(
        totalPages,
        lastPage + (pageCount - (lastPage - firstPage))
      );
    } else {
      firstPage = Math.max(1, firstPage - (pageCount - (lastPage - firstPage)));
    }
  }

  if (lastPage - firstPage + 1 > pageCount) {
    if (currentPage > totalPages / 2) {
      // eslint-disable-next-line
      firstPage++;
    } else {
      // eslint-disable-next-line
      lastPage--;
    }
  }
  const firstResult = limit * (currentPage - 1);
  const lastResult = limit * currentPage - 1;

  return {
    totalPages,
    pages: Math.min(lastPage - firstPage + 1, totalPages),
    currentPage,
    firstPage,
    lastPage,
    previousPage: currentPage - 1,
    nextPage: currentPage + 1,
    hasPreviousPage: currentPage > 1,
    hasNextPage: currentPage < totalPages,
    totalResults,
    results: Math.min(lastResult - firstResult + 1, totalResults),
    firstResult,
    lastResult
  };
}
