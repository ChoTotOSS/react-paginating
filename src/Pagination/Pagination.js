import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { getRange, getPageInfo } from './utils';

class Pagination extends Component {
  constructor() {
    super();
    this.state = {
      currentPage: 0,
    };
  }

  UNSAFE_componentWillMount() {
    if (this.props.currentPage) {
      this.setState({
        currentPage: parseInt(this.props.currentPage, 10),
      });
    }
  }

  UNSAFE_componentWillReceiveProps(nextProps) {
    if (nextProps.currentPage !== this.props.currentPage) {
      this.setState({
        currentPage: parseInt(nextProps.currentPage, 10),
      });
    }
  }

  _getPageItemProps = props => {
    const { pageValue, onPageChange: handlePageChange } = props;

    const onPageChange = () => {
      if (typeof handlePageChange === 'function') {
        handlePageChange(pageValue);
      }
      this.setState({
        currentPage: pageValue,
      });
    };

    return {
      onClick: onPageChange,
    };
  };

  render() {
    const { total, limit, pageCount } = this.props;
    const { currentPage } = this.state;

    const pageInfo = getPageInfo({
      limit,
      pageCount,
      total,
      page: currentPage,
    });

    const {
      firstPage,
      lastPage,
      hasNextPage,
      hasPreviousPage,
      previousPage,
      nextPage,
      totalPages,
    } = pageInfo;

    const pages = total ? getRange(firstPage, lastPage) : [];

    return (
      <div>
        {this.props.children({
          pages,
          previousPage,
          nextPage,
          totalPages,
          currentPage,
          hasNextPage,
          hasPreviousPage,
          getPageItemProps: this._getPageItemProps,
        })}
      </div>
    );
  }
}

Pagination.propTypes = {
  total: PropTypes.number.isRequired,
  limit: PropTypes.number,
  pageCount: PropTypes.number,
  currentPage: PropTypes.number,
  pageValue: PropTypes.number,
  children: PropTypes.func.isRequired,
  onPageChange: PropTypes.func,
};

Pagination.defaultProps = {
  limit: 10,
  pageCount: 5,
  currentPage: 0,
  pageValue: 0,
};

export default Pagination;
