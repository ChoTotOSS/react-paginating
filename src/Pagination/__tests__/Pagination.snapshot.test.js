import React from 'react';
import renderer from 'react-test-renderer';
import Pagination from '../';

const { describe, it } = global;

describe('Pagination render page items', () => {
  it('should render page items 1 -> 5', () => {
    const total = 90;
    const limit = 10;
    const currentPage = 1;
    const tree = renderer
      .create(
        <Pagination
          total={total}
          limit={limit}
          pageCount={5}
          currentPage={currentPage}
        >
          {({
            pages,
            currentPage,
            hasNextPage,
            hasPreviousPage,
            previousPage,
            nextPage,
            totalPages,
            getPageItemProps
          }) => (
            <div>
              <div>
                <a
                  {...getPageItemProps({ pageValue: 1 })}
                  id="page_first"
                  href="1"
                >
                  first
                </a>

                {hasPreviousPage && (
                  <a
                    {...getPageItemProps({
                      pageValue: previousPage
                    })}
                    id="page_prev"
                    href={`${previousPage}`}
                  >
                    {'<'}
                  </a>
                )}

                {pages.map(page => {
                  let activePage = null;
                  if (currentPage === page) {
                    activePage = { backgroundColor: '#fdce09' };
                  }
                  return (
                    <a
                      id={`page_${page}`}
                      key={page}
                      href={`${page}`}
                      {...getPageItemProps({ pageValue: page })}
                    >
                      {page}
                    </a>
                  );
                })}

                {hasNextPage && (
                  <a
                    {...getPageItemProps({ pageValue: nextPage })}
                    id="page_next"
                    href={`${nextPage}`}
                  >
                    {'>'}
                  </a>
                )}

                <a
                  {...getPageItemProps({ pageValue: totalPages })}
                  id="page_last"
                  href={`${totalPages}`}
                >
                  last
                </a>
              </div>
            </div>
          )}
        </Pagination>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render page items 2 -> 6', () => {
    const total = 90;
    const limit = 10;
    const currentPage = 4;
    const tree = renderer
      .create(
        <Pagination
          total={total}
          limit={limit}
          pageCount={5}
          currentPage={currentPage}
        >
          {({
            pages,
            currentPage,
            hasNextPage,
            hasPreviousPage,
            previousPage,
            nextPage,
            totalPages,
            getPageItemProps
          }) => (
            <div>
              <div>
                <a
                  {...getPageItemProps({ pageValue: 1 })}
                  id="page_first"
                  href="1"
                >
                  first
                </a>

                {hasPreviousPage && (
                  <a
                    {...getPageItemProps({
                      pageValue: previousPage
                    })}
                    id="page_prev"
                    href={`${previousPage}`}
                  >
                    {'<'}
                  </a>
                )}

                {pages.map(page => {
                  let activePage = null;
                  if (currentPage === page) {
                    activePage = { backgroundColor: '#fdce09' };
                  }
                  return (
                    <a
                      id={`page_${page}`}
                      key={page}
                      href={`${page}`}
                      {...getPageItemProps({ pageValue: page })}
                    >
                      {page}
                    </a>
                  );
                })}

                {hasNextPage && (
                  <a
                    {...getPageItemProps({ pageValue: nextPage })}
                    id="page_next"
                    href={`${nextPage}`}
                  >
                    {'>'}
                  </a>
                )}

                <a
                  {...getPageItemProps({ pageValue: totalPages })}
                  id="page_last"
                  href={`${totalPages}`}
                >
                  last
                </a>
              </div>
            </div>
          )}
        </Pagination>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
