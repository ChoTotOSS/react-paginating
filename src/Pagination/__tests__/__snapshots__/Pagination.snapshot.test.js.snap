// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination render page items should render page items 1 -> 5 1`] = `
<div>
  <div>
    <div>
      <a
        href="1"
        id="page_first"
        onClick={[Function]}
      >
        first
      </a>
      <a
        href="1"
        id="page_1"
        onClick={[Function]}
      >
        1
      </a>
      <a
        href="2"
        id="page_2"
        onClick={[Function]}
      >
        2
      </a>
      <a
        href="3"
        id="page_3"
        onClick={[Function]}
      >
        3
      </a>
      <a
        href="4"
        id="page_4"
        onClick={[Function]}
      >
        4
      </a>
      <a
        href="5"
        id="page_5"
        onClick={[Function]}
      >
        5
      </a>
      <a
        href="2"
        id="page_next"
        onClick={[Function]}
      >
        &gt;
      </a>
      <a
        href="9"
        id="page_last"
        onClick={[Function]}
      >
        last
      </a>
    </div>
  </div>
</div>
`;

exports[`Pagination render page items should render page items 2 -> 6 1`] = `
<div>
  <div>
    <div>
      <a
        href="1"
        id="page_first"
        onClick={[Function]}
      >
        first
      </a>
      <a
        href="3"
        id="page_prev"
        onClick={[Function]}
      >
        &lt;
      </a>
      <a
        href="2"
        id="page_2"
        onClick={[Function]}
      >
        2
      </a>
      <a
        href="3"
        id="page_3"
        onClick={[Function]}
      >
        3
      </a>
      <a
        href="4"
        id="page_4"
        onClick={[Function]}
      >
        4
      </a>
      <a
        href="5"
        id="page_5"
        onClick={[Function]}
      >
        5
      </a>
      <a
        href="6"
        id="page_6"
        onClick={[Function]}
      >
        6
      </a>
      <a
        href="5"
        id="page_next"
        onClick={[Function]}
      >
        &gt;
      </a>
      <a
        href="9"
        id="page_last"
        onClick={[Function]}
      >
        last
      </a>
    </div>
  </div>
</div>
`;
